src/main/scala/com/microsoft/partnercatalyst/fortis/spark/TransformManager.scala

```diff
@@ -0,0 +1,256 @@
+package com.microsoft.partnercatalyst.fortis.spark
+
+import java.time.Duration
+import java.util.concurrent.locks.ReentrantLock
+import java.util.concurrent.{CompletableFuture, SynchronousQueue, TimeUnit}
+
+import com.microsoft.azure.servicebus.primitives.ConnectionStringBuilder
+import com.microsoft.azure.servicebus._
+import com.microsoft.partnercatalyst.fortis.spark.TransformManager._
+import com.microsoft.partnercatalyst.fortis.spark.dba.ConfigurationManager
+import com.microsoft.partnercatalyst.fortis.spark.dto.{BlacklistedTerm, SiteSettings}
+import com.microsoft.partnercatalyst.fortis.spark.logging.Loggable
+import com.microsoft.partnercatalyst.fortis.spark.transforms.image.{ImageAnalysisAuth, ImageAnalyzer}
+import com.microsoft.partnercatalyst.fortis.spark.transforms.language.{LanguageDetector, LanguageDetectorAuth}
+import com.microsoft.partnercatalyst.fortis.spark.transforms.locations.LocationsExtractorFactory
+import com.microsoft.partnercatalyst.fortis.spark.transforms.locations.client.FeatureServiceClient
+import com.microsoft.partnercatalyst.fortis.spark.transforms.sentiment.SentimentDetectorAuth
+import org.apache.spark.SparkContext
+
+import scala.util.Properties
+
+class TransformManager(configManager: ConfigurationManager, featureServiceClient: FeatureServiceClient) extends Serializable with Loggable {
+  private val shouldUpdate: SynchronousQueue[Delta] = new SynchronousQueue[Delta]()
+  private val writeLock: ReentrantLock = new ReentrantLock(true)
+
+  // Do not serialize these values. The transformContext would otherwise contain stale Broadcast
+  // instances, so we rebuild it on recovery from checkpoint.
+  @volatile @transient private var transformContext: TransformContext = null
+  @volatile @transient private var queueClient: QueueClient = null
+
+  def getOrUpdateContext(sparkContext: SparkContext): TransformContext = {
+    ensureInitialized(sparkContext)
+
+    writeLock.lock()
+    try {
+      // Grab delta from hand-off only if it's available *now*
+      val delta = Option(shouldUpdate.poll(0, TimeUnit.SECONDS))
+
+      // Update transform context with delta
+      delta.foreach(updateTransformContext(_, sparkContext))
+    }
+    finally {
+      writeLock.unlock()
+    }
+
+    transformContext
+  }
+
+  private def ensureInitialized(sparkContext: SparkContext): Unit = {
+    if (transformContext == null) {
+      // Blocking init of transform context and queue client used for non-blocking updates.
+      // Initialization has not yet occurred since construction/deserialization.
+      writeLock.lock()
+      try {
+        if (transformContext == null) {
```

Comment:
This is essentially implementing double check locking which is easily broken in many memory models (sample [article for .NET](http://csharpindepth.com/Articles/General/Singleton.aspx#dcl)), similar resources exist for [the JVM](https://stackoverflow.com/a/25262497/3817588). It would be simpler and safer to use a `lazy` field instead.